File: test/test_unique.cpp
#include <stdio.h>
#include "TRTCContext.h"
#include "DVVector.h"
#include "unique.h"


int main()
{
	TRTCContext ctx;
	{
		int h_value[7] = { 1, 3, 3, 3, 2, 2, 1 };
		DVVector d_value(ctx, "int32_t", 7, h_value);
		int count = TRTC_Unique(ctx, d_value);
		d_value.to_host(h_value, 0, count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_value[i]);
		puts("");
	}

	Functor equal(ctx, {}, { "x", "y" }, "        return x==y;\n");

	{
		int h_value[7] = { 1, 3, 3, 3, 2, 2, 1 };
		DVVector d_value(ctx, "int32_t", 7, h_value);
		int count = TRTC_Unique(ctx, d_value, equal);
		d_value.to_host(h_value, 0, count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_value[i]);
		puts("");
	}

	{
		int h_in[7] = { 1, 3, 3, 3, 2, 2, 1 };
		DVVector d_in(ctx, "int32_t", 7, h_in);
		int h_out[7];
		DVVector d_out(ctx, "int32_t", 7);
		int count = TRTC_Unique_Copy(ctx, d_in, d_out);
		d_out.to_host(h_out, 0, count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_out[i]);
		puts("");
	}

	{
		int h_in[7] = { 1, 3, 3, 3, 2, 2, 1 };
		DVVector d_in(ctx, "int32_t", 7, h_in);
		int h_out[7];
		DVVector d_out(ctx, "int32_t", 7);
		int count = TRTC_Unique_Copy(ctx, d_in, d_out, equal);
		d_out.to_host(h_out, 0, count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_out[i]);
		puts("");
	}

	{
		int h_keys[7] = { 1, 3, 3, 3, 2, 2, 1 };
		int h_values[7] = { 9, 8, 7, 6, 5, 4, 3 };
		DVVector d_keys(ctx, "int32_t", 7, h_keys);
		DVVector d_values(ctx, "int32_t", 7, h_values);
		int count = TRTC_Unique_By_Key(ctx, d_keys, d_values);
		d_keys.to_host(h_keys, 0, count);
		d_values.to_host(h_values, 0, count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_keys[i]);
		puts("");
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_values[i]);
		puts("");
	}

	{
		int h_keys[7] = { 1, 3, 3, 3, 2, 2, 1 };
		int h_values[7] = { 9, 8, 7, 6, 5, 4, 3 };
		DVVector d_keys(ctx, "int32_t", 7, h_keys);
		DVVector d_values(ctx, "int32_t", 7, h_values);
		int count = TRTC_Unique_By_Key(ctx, d_keys, d_values, equal);
		d_keys.to_host(h_keys, 0, count);
		d_values.to_host(h_values, 0, count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_keys[i]);
		puts("");
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_values[i]);
		puts("");
	}

	{
		int h_keys_in[7] = { 1, 3, 3, 3, 2, 2, 1 };
		int h_values_in[7] = { 9, 8, 7, 6, 5, 4, 3 };
		DVVector d_keys_in(ctx, "int32_t", 7, h_keys_in);
		DVVector d_values_in(ctx, "int32_t", 7, h_values_in);

		int h_keys_out[7];
		int h_values_out[7];
		DVVector d_keys_out(ctx, "int32_t", 7);
		DVVector d_values_out(ctx, "int32_t", 7);

		int count = TRTC_Unique_By_Key_Copy(ctx, d_keys_in, d_values_in, d_keys_out, d_values_out);
		d_keys_out.to_host(h_keys_out, 0, count);
		d_values_out.to_host(h_values_out, 0, count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_keys_out[i]);
		puts("");
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_values_out[i]);
		puts("");
	}

	{
		int h_keys_in[7] = { 1, 3, 3, 3, 2, 2, 1 };
		int h_values_in[7] = { 9, 8, 7, 6, 5, 4, 3 };
		DVVector d_keys_in(ctx, "int32_t", 7, h_keys_in);
		DVVector d_values_in(ctx, "int32_t", 7, h_values_in);

		int h_keys_out[7];
		int h_values_out[7];
		DVVector d_keys_out(ctx, "int32_t", 7);
		DVVector d_values_out(ctx, "int32_t", 7);

		int count = TRTC_Unique_By_Key_Copy(ctx, d_keys_in, d_values_in, d_keys_out, d_values_out, equal);
		d_keys_out.to_host(h_keys_out, 0, count);
		d_values_out.to_host(h_values_out, 0, count);
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_keys_out[i]);
		puts("");
		for (uint32_t i = 0; i < count; i++)
			printf("%d ", h_values_out[i]);
		puts("");
	}

	return 0;
}

File: CSharp/test/test_unique/test_unique.cs
using System;
using ThrustRTCSharp;

namespace test_unique
{
    class test_unique
    {
        static void Main(string[] args)
        {
            {
                DVVector d_value = new DVVector(new int[] { 1, 3, 3, 3, 2, 2, 1 });
                long count = TRTC.Unique(d_value);
                print_array((int[])d_value.to_host(0, count));
            }

            {
                DVVector d_value = new DVVector(new int[] { 1, 3, 3, 3, 2, 2, 1 });
                long count = TRTC.Unique(d_value, new Functor("EqualTo"));
                print_array((int[])d_value.to_host(0, count));
            }

            {
                DVVector d_in = new DVVector(new int[] { 1, 3, 3, 3, 2, 2, 1 });
                DVVector d_out = new DVVector("int32_t", 7);
                long count = TRTC.Unique_Copy(d_in, d_out);
                print_array((int[])d_out.to_host(0, count));
            }

            {
                DVVector d_in = new DVVector(new int[] { 1, 3, 3, 3, 2, 2, 1 });
                DVVector d_out = new DVVector("int32_t", 7);
                long count = TRTC.Unique_Copy(d_in, d_out, new Functor("EqualTo"));
                print_array((int[])d_out.to_host(0, count));
            }

            {
                DVVector d_keys = new DVVector(new int[] { 1, 3, 3, 3, 2, 2, 1 });
                DVVector d_values = new DVVector(new int[] { 9, 8, 7, 6, 5, 4, 3 });
                long count = TRTC.Unique_By_Key(d_keys, d_values);
                print_array((int[])d_keys.to_host(0, count));
                print_array((int[])d_values.to_host(0, count));
            }

            {
                DVVector d_keys = new DVVector(new int[] { 1, 3, 3, 3, 2, 2, 1 });
                DVVector d_values = new DVVector(new int[] { 9, 8, 7, 6, 5, 4, 3 });
                long count = TRTC.Unique_By_Key(d_keys, d_values, new Functor("EqualTo"));
                print_array((int[])d_keys.to_host(0, count));
                print_array((int[])d_values.to_host(0, count));
            }

            {
                DVVector d_keys_in = new DVVector(new int[] { 1, 3, 3, 3, 2, 2, 1 });
                DVVector d_values_in = new DVVector(new int[] { 9, 8, 7, 6, 5, 4, 3 });
                DVVector d_keys_out = new DVVector("int32_t", 7);
                DVVector d_values_out = new DVVector("int32_t", 7);
                long count = TRTC.Unique_By_Key_Copy(d_keys_in, d_values_in, d_keys_out, d_values_out);
                print_array((int[])d_keys_out.to_host(0, count));
                print_array((int[])d_values_out.to_host(0, count));
            }

            {
                DVVector d_keys_in = new DVVector(new int[] { 1, 3, 3, 3, 2, 2, 1 });
                DVVector d_values_in = new DVVector(new int[] { 9, 8, 7, 6, 5, 4, 3 });
                DVVector d_keys_out = new DVVector("int32_t", 7);
                DVVector d_values_out = new DVVector("int32_t", 7);
                long count = TRTC.Unique_By_Key_Copy(d_keys_in, d_values_in, d_keys_out, d_values_out, new Functor("EqualTo"));
                print_array((int[])d_keys_out.to_host(0, count));
                print_array((int[])d_values_out.to_host(0, count));
            }
        }

        static void print_array<T>(T[] arr)
        {
            foreach (var item in arr)
            {
                Console.Write(item.ToString());
                Console.Write(" ");
            }
            Console.WriteLine("");
        }
    }
}
